stellargraph/mapper/graphwave_generator.py

```diff
@@ -0,0 +1,238 @@
+                "a graph with multiple node types is passed. Stopping.".format(
+                    type(self).__name__
+                )
+            )
+
+        # Create sparse adjacency matrix:
+        adj = G.to_adjacency_matrix().tocoo()
+
+        # Function to map node IDs to indices for quicker node index lookups
+        self._node_lookup = G._get_index_for_nodes
+
+        degree_mat = diags(np.asarray(adj.sum(1)).ravel())
+        laplacian = degree_mat - adj
+
+        # TODO: add in option to compute wavelet transform using Chebysev polynomials
+
+        self.scales = np.array(scales).astype(np.float32)
+
+        if num_eigenvecs == -1:
+            eigen_vals, eigen_vecs = self.sufficiently_sampled_eigs(
+                laplacian, self.scales.min()
+            )
+        else:
+            eigen_vals, eigen_vecs = eigs(laplacian, k=num_eigenvecs, sigma=1e-3)
+            eigen_vals = np.real(eigen_vals).astype(np.float32)
+            eigen_vecs = eigen_vecs[:, 1:]
+            eigen_vals = eigen_vals[1:]
+
+        self.eigen_vecs = np.real(eigen_vecs).astype(np.float32)
+
+        # the columns of U exp(-scale * eigenvalues) U^T (U = eigenvectors) are used to calculate the node embeddings
+        # (each column corresponds to a node)
+        # to avoid computing a dense NxN matrix when only several eigenvalues are specified
+        # U exp(-scale * eigenvalues) is computed and stored - which is an N x num_eigenvectors matrix
+        # the columns of U exp(-scale * eigenvalues) U^T are then computed on the fly in generator.flow()
+
+        # a list of [U exp(-scale * eigenvalues) for scale in scales]
+        Ues = [
+            (self.eigen_vecs * np.exp(-s * eigen_vals))[np.newaxis, :, :].astype(
+                np.float32
+            )
+            for s in scales
+        ]
+        self.Ues = tf.convert_to_tensor(np.concatenate(Ues, axis=0))
+
+    @staticmethod
+    def sufficiently_sampled_eigs(laplacian, min_scale):
+        """
+        This function calculates increasing numbers of eigenvalues using a binary search until a sufficient number of
+        eigenvalues have been found to ensure accurate results of the GraphWave algorithm.
+
+        Args:
+            laplacian: the un-normalized graph laplacian.
+            min_scale: the smallest wavelet scale used.
+        Returns:
+            eigen_vals: the eigenvalues found.
+            eigen_vecs: the eigenvectors found.
+        """
+        max_num_eigs = laplacian.shape[0] - 2
+
+        prev_eig_l2_norm = 0.0
+        eig_l2_norm = -1
+        prev_eig_max = 1e-3
+        k = min(16, max_num_eigs)
+
+        eigen_vals = np.array([])
+        eigen_vecs = np.empty((laplacian.shape[0], 0), dtype=np.float32)
+
+        # use increasing k (doubling per iter) until the filtered eigenvalue l2 norm
+        # increases by less than 10% since the last iter.
+        while (k <= max_num_eigs) and ((prev_eig_l2_norm / eig_l2_norm) < 0.9):
```

Comment:
Finally, why is 10% an appropriate threshold?

Reply:
Somewhat of a magic number. It seems to be sufficient for cora + if doubling (roughly...) the number of eigenvalues only increases the norm by 10% and the contribution of subsequent eigenvalues decreases exponentially then it - seems - safe to stop here. 

We could make this a user param here - I guess its like a solver coming across a local extrema and stopping once the difference between steps decreases below a certain threshold.